source/hwIo.py

```diff
@@ -89,18 +91,24 @@ def waitForRead(self, timeout):
 				timeout -= int((time.time()-curTime)*1000)
 
 	def write(self, data):
+		"""
+		@type data: bytes
+		"""
+		if not isinstance(data, bytes):

```

Comment:
I think we should use collections.abc.ByteString here, so that bytearray is also allowed as a valid type. It is fully compatible.

Reply:
Perhaps, but I think it's outside of the scope of this PR to go down that road. EG finding the optimal types to allow the greatest flexibility. This PR should just be concerned with getting it working. If someone shows that it causes some perf concern to do `bytes(myByteArray)`, or taking a `bytes` type only is too inflexible for some given case we can look at changing it.

Reply:
Note that before this pr, the write function just accepted everything you threwat it. You're now introducing this explicit check for bytes. I won't insist, but I really believe that it is too strict, especially given bytearray is fully compatible with bytes and they are both handled by the ByteString abstract class.

Reply:
I reduced the types it could take to simplify verifying correctness. Taking multiple types introduces uncertainty, for instance taking an `int`, the code must guess how to convert it (number of bytes, byte order, signed), these are device specific and should be decided in the driver. There is certainly an argument for accepting a `bytearray`, however I don't see that it's necessary, and doesn't greatly impact the code. While it's tempting to try to build a general purpose library, it's much easier to get the known cases correct if we focus on those only.

> especially given bytearray is fully compatible with bytes 

It should be noted that a `bytearray` is not convertible by `ctypes.create_string_buffer` and will not be automatically converted to a bytes object.